tests/run-pass/concurrency/sync.rs

```diff
@@ -0,0 +1,274 @@
+        let thread = thread::spawn(move || {
+            let mut data = data.lock().unwrap();
+            thread::yield_now();
+            *data += 1;
+        });
+        threads.push(thread);
+    }
+
+    for thread in threads {
+        thread.join().unwrap();
+    }
+
+    assert!(data.try_lock().is_ok());
+
+    let data = Arc::try_unwrap(data).unwrap().into_inner().unwrap();
+    assert_eq!(data, 3);
+}
+
+fn check_rwlock_write() {
+    let data = Arc::new(RwLock::new(0));
+    let mut threads = Vec::new();
+
+    for _ in 0..3 {
+        let data = Arc::clone(&data);
+        let thread = thread::spawn(move || {
+            let mut data = data.write().unwrap();
+            thread::yield_now();
+            *data += 1;
+        });
+        threads.push(thread);
+    }
+
+    for thread in threads {
+        thread.join().unwrap();
+    }
+
+    assert!(data.try_write().is_ok());
+
+    let data = Arc::try_unwrap(data).unwrap().into_inner().unwrap();
+    assert_eq!(data, 3);
+}
+
+fn check_rwlock_read_no_deadlock() {
+    let l1 = Arc::new(RwLock::new(0));
+    let l2 = Arc::new(RwLock::new(0));
+
+    let l1_copy = Arc::clone(&l1);
+    let l2_copy = Arc::clone(&l2);
+    let _guard1 = l1.read().unwrap();
+    let handle = thread::spawn(move || {
+        let _guard2 = l2_copy.read().unwrap();
+        thread::yield_now();
+        let _guard1 = l1_copy.read().unwrap();
+    });
+    thread::yield_now();
+    let _guard2 = l2.read().unwrap();
+    handle.join().unwrap();
+}
+
+// Check if channels are working.
+
+/// The test taken from the Rust documentation.
+fn simple_send() {
+    let (tx, rx) = channel();
+    thread::spawn(move || {
+        tx.send(10).unwrap();
+    });
+    assert_eq!(rx.recv().unwrap(), 10);
+}
+
+/// The test taken from the Rust documentation.
+fn multiple_send() {
+    let (tx, rx) = channel();
+    for i in 0..10 {
+        let tx = tx.clone();
+        thread::spawn(move || {
+            tx.send(i).unwrap();
+        });
+    }
+
+    let mut sum = 0;
+    for _ in 0..10 {
+        let j = rx.recv().unwrap();
+        assert!(0 <= j && j < 10);
+        sum += j;
+    }
+    assert_eq!(sum, 45);
+}
+
+/// The test taken from the Rust documentation.
+fn send_on_sync() {
+    let (sender, receiver) = sync_channel(1);
+
+    // this returns immediately
+    sender.send(1).unwrap();
+
+    thread::spawn(move || {
+        // this will block until the previous message has been received
+        sender.send(2).unwrap();
+    });
+
+    assert_eq!(receiver.recv().unwrap(), 1);
+    assert_eq!(receiver.recv().unwrap(), 2);
+}
+
+// Check if Rust once statics are working.
+
+static mut VAL: usize = 0;
+static INIT: Once = Once::new();
+
+fn get_cached_val() -> usize {
+    unsafe {
+        INIT.call_once(|| {
+            VAL = expensive_computation();
+        });
+        VAL
+    }
+}
+
+fn expensive_computation() -> usize {
+    let mut i = 1;
+    let mut c = 1;
+    while i < 10000 {
```

Comment:
10k is a *lot* for Miri... how long does this test take to run?

Reply:
Does not feel problematic:
```
> time ./miri test tests/run-pass/concurrency/sync.rs
...
0.42user 0.27system 0:00.69elapsed 100%CPU (0avgtext+0avgdata 87052maxresident)k
0inputs+576outputs (0major+86990minor)pagefaults 0swaps
```

Reply:
That's 0.7s if I made the right assumptions about units? Maybe halve the iteration count then to reduce it to less than 0.5s. I don't see a good reason to make it take longer than that.

Reply:
Reducing from 10000 to 1000 gives no measurable difference in running time.

Reply:
Oh... that's odd. I wonder where it spends its time then.

Not too big of a deal though.